appcues/src/main/java/com/appcues/util/NotificationBuilderExt.kt

```diff
@@ -0,0 +1,99 @@
+package com.appcues.util
+
+import android.annotation.SuppressLint
+import android.app.Application
+import android.app.NotificationChannel
+import android.app.NotificationManager
+import android.app.PendingIntent
+import android.content.Context
+import android.content.Intent
+import android.graphics.Bitmap
+import android.graphics.BitmapFactory
+import android.net.Uri
+import android.os.Build
+import androidx.core.app.NotificationCompat
+import androidx.core.app.NotificationManagerCompat
+import androidx.core.content.ContextCompat
+import com.appcues.AppcuesFirebaseMessagingService.AppcuesMessagingData
+import com.appcues.R
+import com.google.firebase.messaging.RemoteMessage
+import java.io.InputStream
+import java.net.HttpURLConnection
+import java.net.URL
+
+// for now channels are not configurable, which means all push messages will come through the default channel defined in XML config
+internal fun Context.getNotificationBuilder(): NotificationCompat.Builder {
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+        with(getSystemService(Application.NOTIFICATION_SERVICE) as NotificationManager) {
+
+            createNotificationChannel(
+                NotificationChannel(
+                    getString(R.string.appcues_notification_channel_id),
+                    getString(R.string.appcues_notification_channel_name),
+                    resources.getInteger(R.integer.appcues_notification_channel_importance)
+                ).apply { description = getString(R.string.appcues_notification_channel_description) }
+            )
+        }
+    }
+
+    return NotificationCompat.Builder(this, getString(R.string.appcues_notification_channel_id))
+}
+
+internal fun NotificationCompat.Builder.setupNotification(message: RemoteMessage) = apply {
+    setAutoCancel(true)
+    setPriority(message.priority)
+    message.notification?.visibility?.let { setVisibility(it) }
+    message.notification?.channelId?.let { setChannelId(it) }
+}
+
+internal fun NotificationCompat.Builder.setContent(data: AppcuesMessagingData) = apply {
+    setContentTitle(data.title)
+    setContentText(data.body)
+}
+
+internal fun NotificationCompat.Builder.setStyle(context: Context, data: AppcuesMessagingData) = apply {
+    downloadImageFromUrl(data.image)?.let { image ->
+        setStyle(
+            NotificationCompat.BigPictureStyle()
+                .bigPicture(image)
+                .setSummaryText(data.body)
+        )
+    } ?: run {
+        setStyle(
+            NotificationCompat.BigTextStyle()
+                .bigText(data.body)
+        )
+    }
+
+    setSmallIcon(R.drawable.appcues_notification_small_icon)
+    setColor(ContextCompat.getColor(context, R.color.appcues_notification_color))
```

Comment:
should these be here? or should they use the host application defaults?

Reply:
so this is the "default" value  that they can override in their XML, but I think this is a nice setup, if they dont do anything it will just work with our default values (appcues_color set to 5C5CFF) and the small notification icon, both of which they can replace with their own.

I made it this way because there is no good way for me to get a valid icon, or even the color to set automatically as default